clientv3/grpc_dial_test.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2017 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clientv3
+
+import (
+	"net"
+	"testing"
+	"time"
+
+	"github.com/coreos/etcd/pkg/testutil"
+
+	"sync"
+
+	"google.golang.org/grpc"
+)
+
+// TestBalancerDoNotBlockOnClose ensures that balancer and grpc don't deadlock each other
+// due to rapid open/close conn. The deadlock causes balancer.Close() to block forever.
+// See issue: https://github.com/coreos/etcd/issues/7283 for more detail.
+func TestBalancerDoNotBlockOnClose(t *testing.T) {
+	defer testutil.AfterTest(t)
+
+	kcl := NewKillConnListener([]string{"127.0.0.1:8081", "127.0.0.1:8082"})
+	kcl.Start()
+	defer kcl.Close()
+
+	timer := time.After(10 * time.Second)
```

Comment:
From testing, the old balancer doesn't get block immediately. Sometime it can take up to 2 seconds to trigger the deadlock. So I use 10 seconds to be safe.